yarn-project/p2p/src/client/p2p_client.ts

```diff
@@ -58,118 +58,144 @@ export interface P2PSyncState {
-     * @returns A boolean flag indicating readiness.
-     */
-    isReady(): boolean;
-
-    /**
-     * Returns the current status of the p2p client.
-     */
-    getStatus(): Promise<P2PSyncState>;
-
-    /**
-     * Returns the ENR of this node, if any.
-     */
-    getEnr(): ENR | undefined;
-
-    /** Identifies a p2p client. */
-    isP2PClient(): true;
-
-    updateP2PConfig(config: Partial<P2PConfig>): Promise<void>;
-  };
+export type P2P<T extends P2PClientType = P2PClientType.Full> = P2PApi<T> & {
+  /**
+   * Broadcasts a block proposal to other peers.
+   *
+   * @param proposal - the block proposal
+   */
+  broadcastProposal(proposal: BlockProposal): void;
+
+  /**
+   * Registers a callback from the validator client that determines how to behave when
+   * foreign block proposals are received
+   *
+   * @param handler - A function taking a received block proposal and producing an attestation
+   */
+  // REVIEW: https://github.com/AztecProtocol/aztec-packages/issues/7963
+  // ^ This pattern is not my favorite (md)
+  registerBlockProposalHandler(handler: (block: BlockProposal) => Promise<BlockAttestation | undefined>): void;
+
+  /**
+   * Request a list of transactions from another peer by their tx hashes.
+   * @param txHashes - Hashes of the txs to query.
+   * @returns A list of transactions or undefined if the transactions are not found.
+   */
+  requestTxs(txHashes: TxHash[]): Promise<(Tx | undefined)[]>;
+
+  /**
+   * Request a transaction from another peer by its tx hash.
+   * @param txHash - Hash of the tx to query.
+   */
+  requestTxByHash(txHash: TxHash): Promise<Tx | undefined>;
+
+  /**
+   * Verifies the 'tx' and, if valid, adds it to local tx pool and forwards it to other peers.
+   * @param tx - The transaction.
+   **/
+  sendTx(tx: Tx): Promise<void>;
+
+  /**
+   * Adds transactions to the pool. Does not send to peers or validate the tx.
+   * @param txs - The transactions.
+   **/
+  addTxs(txs: Tx[]): Promise<void>;
+
+  /**
+   * Deletes 'txs' from the pool, given hashes.
+   * NOT used if we use sendTx as reconcileTxPool will handle this.
+   * @param txHashes - Hashes to check.
+   **/
+  deleteTxs(txHashes: TxHash[]): Promise<void>;
+
+  /**
+   * Returns a transaction in the transaction pool by its hash.
+   * @param txHash  - Hash of tx to return.
+   * @returns A single tx or undefined.
+   */
+  getTxByHashFromPool(txHash: TxHash): Promise<Tx | undefined>;
+
+  /**
+   * Returns transactions in the transaction pool by hash.
+   * @param txHashes  - Hashes of txs to return.
+   * @returns An array of txs or undefined.
+   */
+  getTxsByHashFromPool(txHashes: TxHash[]): Promise<(Tx | undefined)[]>;
+
+  /**
+   * Checks if transactions exist in the pool and returns the hashes of those that don't
+   * @param txHashes - The hashes of the transactions to check for
+   * @returns The transaction hashes if not found
+   */
+  getUnavailableTxsFromPool(txHashes: TxHash[]): Promise<TxHash[]>;
```

Comment:
Nit: it feels odd to have a "negative" method in the API. Maybe change to `hasTxs(txHashes): Promise<boolean[]>` instead, which is the first step of what the implementation of this method does? Again, just cosmetic, and no big deal given it's just internal API.

Reply:
Done